corehq/apps/export/multithreaded.py

```diff
@@ -0,0 +1,379 @@
+
+    def get_result(self):
+        return RetryResult(self.page, self.path, self.page_size, 0)
+
+
+def rebuild_export_mutithreaded(export_id, num_processes, page_size=100000):
+    assert num_processes > 0
+
+    def _log_page_dumped(paginator):
+        logger.info('  Dump page {} complete: {} docs'.format(paginator.page, paginator.page_size))
+
+    export_instance = get_properly_wrapped_export_instance(export_id)
+    filters = export_instance.get_filters()
+    total_docs = get_export_size(export_instance, filters)
+
+    logger.info('Starting data dump of {} docs'.format(total_docs))
+    exporter = MultithreadedExporter(export_instance, total_docs, num_processes)
+    paginator = OutputPaginator(export_id)
+    with exporter:
+        with paginator:
+            for index, doc in enumerate(_get_export_documents(export_instance, filters)):
+                paginator.write(doc)
+                if paginator.page_size == page_size:
+                    _log_page_dumped(paginator)
+                    exporter.process_page(paginator.get_result())
+                    paginator.next_page()
+            if paginator.page_size:
+                _log_page_dumped(paginator)
+                exporter.process_page(paginator.get_result())
+
+        exporter.wait_till_completion()
+
+
+def run_export_with_logging(export_instance, page_number, dump_path, doc_count, attempts):
+    """Log any exceptions here since logging on the other side of the process queue
+    won't show the traceback
+    """
+    logger.info('    Processing page {} started (attempt {})'.format(page_number, attempts))
+    progress_queue = getattr(run_export, 'queue', None)
+    update_frequency = min(1000, int(doc_count / 10) or 1)
+    progress_tracker = LoggingProgressTracker(page_number, progress_queue, update_frequency)
+    try:
+        result = run_export(export_instance, page_number, dump_path, doc_count, progress_tracker)
+        if progress_queue:
+            # just to make sure we set progress to 100%
+            progress_queue.put(ProgressValue(page_number, doc_count, doc_count))
+        logger.info('    Processing page {} complete'.format(page_number))
+        return result
+    except Exception:
+        logger.exception("Error processing page {} (attempt {})".format(page_number, attempts))
+        raise
+
+
+def run_export(export_instance, page_number, dump_path, doc_count, progress_tracker=None):
+    docs = _get_export_documents_from_file(dump_path, doc_count)
+    export_file = get_export_file(export_instance, docs, progress_tracker)
+    return SuccessResult(page_number, export_file.path, doc_count)
+
+
+def _get_export_documents_from_file(dump_path, doc_count):
+    """Mimic the results of an ES scroll query but get results from jsonlines file"""
+    def _doc_iter():
+        with gzip.open(dump_path) as f:
+            for line in f.readlines():
+                yield json.loads(line)
+        os.remove(dump_path)
```

Comment:
this line should probably be in a `finally` clause in case an exception is raised during iteration, and the above `with` block should be in the `try` block.

Reply:
actually no since they get added to the final ZIP so that we can retry them later. But I did notice that they weren't removed after that so have removed them now: ab8ab17